File: apps/ios/Voiid/Voiid/Main/ClipFullscreenView.swift
//
//  ClipFullscreenView.swift
//  Voiid
//
//  Instagram-Reels-style fullscreen player. Tap comments/caption -> the reel
//  shrinks into a box at the top and the comments panel slides up below it.
//

import SwiftUI

struct ClipFullscreenView: View {
    let clip: VClip
    @EnvironmentObject var clips: ClipsStore
    @Environment(\.dismiss) private var dismiss
    @State private var liked = false
    @State private var showComments = false
    @State private var commentDraft = ""

    var body: some View {
        GeometryReader { geo in
            VStack(spacing: 0) {
                // The reel — full screen normally; shrinks to a top box when comments open.
                reel
                    .frame(height: showComments ? geo.size.height * 0.42 : geo.size.height)
                    .clipped()

                if showComments {
                    commentsPanel
                        .frame(height: geo.size.height * 0.58)
                        .transition(.move(edge: .bottom))
                }
            }
            .animation(.spring(response: 0.4, dampingFraction: 0.85), value: showComments)
        }
        .background(Color.black.ignoresSafeArea())
        .ignoresSafeArea(.keyboard)
    }

    // MARK: reel

    private var reel: some View {
        ZStack {
            LinearGradient(colors: [VoiidColor.primary, .black], startPoint: .top, endPoint: .bottom)
            Image(systemName: "play.circle.fill").font(.system(size: 64)).foregroundColor(.white.opacity(0.85))

            VStack {
                HStack {
                    Button { showComments ? closeComments() : dismiss() } label: {
                        Image(systemName: showComments ? "chevron.down" : "chevron.left")
                            .font(.title2).foregroundColor(.white)
                    }
                    Spacer()
                }
                .padding()
                Spacer()
                HStack(alignment: .bottom) {
                    VStack(alignment: .leading, spacing: VoiidSpacing.sm) {
                        HStack { VoiidAvatar(size: 44).clipShape(Circle())
                            Text(clip.authorName).font(VoiidFont.rounded(15, .semibold)).foregroundColor(.white) }
                        if !showComments {
                            Text(clip.heading).font(VoiidFont.rounded(14, .regular)).foregroundColor(.white)
                            // Tapping the caption opens comments (per request)
                            Button { openComments() } label: {
                                Text(clip.caption).font(VoiidFont.rounded(12, .regular))
                                    .foregroundColor(.white.opacity(0.85)).underline()
                            }
                        }
                    }
                    Spacer()
                    if !showComments {
                        VStack(spacing: VoiidSpacing.lg) {
                            action(liked ? "heart.fill" : "heart", "\(clip.likes + (liked ? 1 : 0))",
                                   liked ? VoiidColor.error : .white) {
                                Haptics.tap(); liked.toggle(); if liked { clips.toggleLike(clip) }
                            }
                            action("bubble.right.fill", "\(clip.comments)", .white) { openComments() }
                            action("paperplane.fill", "Share", .white) {}
                        }
                    }
                }
                .padding()
            }
        }
    }

    // MARK: comments panel (slides up below the shrunken reel)

    private var commentsPanel: some View {
        VStack(spacing: 0) {
            Capsule().fill(VoiidColor.divider).frame(width: 40, height: 4).padding(.vertical, VoiidSpacing.sm)
            HStack {
                Text("\(clips.comments.count) comments")
                    .font(VoiidFont.rounded(15, .semibold)).foregroundColor(VoiidColor.textPrimary)
                Spacer()
                Button { closeComments() } label: {
                    Image(systemName: "xmark").foregroundColor(VoiidColor.textSecondary)
                }
            }
            .padding(.horizontal, VoiidSpacing.lg).padding(.bottom, VoiidSpacing.sm)
            Divider()

            ScrollView {
                LazyVStack(alignment: .leading, spacing: VoiidSpacing.md) {
                    ForEach(clips.comments) { c in
                        HStack(alignment: .top, spacing: VoiidSpacing.sm) {
                            VoiidAvatar(size: 32, imageName: c.authorPhoto).clipShape(Circle())
                            VStack(alignment: .leading, spacing: 2) {
                                Text(c.authorName).font(VoiidFont.rounded(13, .semibold)).foregroundColor(VoiidColor.textPrimary)
                                Text(c.text).font(VoiidFont.rounded(14, .regular)).foregroundColor(VoiidColor.textPrimary)
                            }
                            Spacer()
                            Image(systemName: "heart").font(.system(size: 13)).foregroundColor(VoiidColor.textSecondary)
                        }
                    }
                }
                .padding(VoiidSpacing.lg)
            }

            Divider()
            HStack(spacing: VoiidSpacing.sm) {
                TextField("", text: $commentDraft,
                          prompt: Text("Add a comment…").foregroundColor(VoiidColor.placeholder))
                    .font(VoiidFont.rounded(15, .regular)).foregroundColor(VoiidColor.textPrimary)
                    .padding(.horizontal, VoiidSpacing.md).frame(height: 44)
                    .background(VoiidColor.fieldFill).clipShape(Capsule())
                Button {
                    let t = commentDraft.trimmingCharacters(in: .whitespaces); guard !t.isEmpty else { return }
                    Haptics.tap(); clips.addComment(t); commentDraft = ""
                } label: {
                    Image(systemName: "paperplane.fill").font(.system(size: 22)).foregroundColor(VoiidColor.primary)
                }
            }
            .padding(.horizontal, VoiidSpacing.md).padding(.vertical, VoiidSpacing.sm)
        }
        .background(VoiidColor.background)
        .clipShape(RoundedRectangle(cornerRadius: 20, style: .continuous))
    }

    private func openComments() { Haptics.tap(); withAnimation { showComments = true } }
    private func closeComments() { withAnimation { showComments = false } }

    private func action(_ icon: String, _ label: String, _ color: Color, _ tap: @escaping () -> Void) -> some View {
        Button(action: tap) {
            VStack(spacing: 4) {
                Image(systemName: icon).font(.system(size: 26)).foregroundColor(color)
                Text(label).font(VoiidFont.rounded(11, .medium)).foregroundColor(.white)
            }
        }
    }
}

File: apps/android/app/src/main/java/com/voiid/app/main/ClipFullscreenView.kt
package com.voiid.app.main

import androidx.activity.compose.BackHandler
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material.icons.automirrored.outlined.Chat
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.ChevronLeft
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.KeyboardArrowDown
import androidx.compose.material.icons.filled.PlayCircle
import androidx.compose.material.icons.outlined.FavoriteBorder
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import com.voiid.app.model.ClipsStore
import com.voiid.app.model.VClip
import com.voiid.app.model.VClipComment
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.components.VoiidAvatar
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont

/** Instagram-Reels-style fullscreen player with inline shrink-to-comments (iOS `ClipFullscreenView`). */
@Composable
fun ClipFullscreenView(clip: VClip, clips: ClipsStore, onClose: () -> Unit) {
    val haptics = LocalVoiidHaptics.current
    var liked by remember { mutableStateOf(false) }
    var showComments by remember { mutableStateOf(false) }
    var commentDraft by remember { mutableStateOf("") }

    BackHandler { if (showComments) showComments = false else onClose() }

    BoxWithConstraints(Modifier.fillMaxSize().background(Color.Black)) {
        val totalH = maxHeight
        val reelFraction by animateFloatAsState(
            targetValue = if (showComments) 0.42f else 1f,
            animationSpec = spring(dampingRatio = 0.85f, stiffness = Spring.StiffnessLow),
            label = "reelHeight",
        )
        Column(Modifier.fillMaxSize()) {
            // Reel — full height normally; shrinks to a top box when comments open.
            Box(Modifier.fillMaxWidth().height(totalH * reelFraction)) {
                Box(
                    Modifier.fillMaxSize().background(Brush.verticalGradient(listOf(VoiidColor.primary, Color.Black))),
                    contentAlignment = Alignment.Center,
                ) {
                    Icon(Icons.Default.PlayCircle, null, tint = Color.White.copy(alpha = 0.85f), modifier = Modifier.size(64.dp))
                }
                // top bar
                Row(Modifier.fillMaxWidth().statusBarsPadding().padding(16.dp)) {
                    Icon(
                        if (showComments) Icons.Default.KeyboardArrowDown else Icons.Default.ChevronLeft, "Back",
                        tint = Color.White,
                        modifier = Modifier.size(28.dp).clickable { if (showComments) showComments = false else onClose() },
                    )
                }
                // bottom author + caption + actions (hidden when comments open)
                if (!showComments) {
                    Row(
                        Modifier.fillMaxWidth().align(Alignment.BottomStart).navigationBarsPadding().padding(16.dp),
                        verticalAlignment = Alignment.Bottom,
                    ) {
                        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                                VoiidAvatar(size = 44.dp, modifier = Modifier.clip(CircleShape))
                                Text(clip.authorName, style = VoiidFont.rounded(15, FontWeight.SemiBold), color = Color.White)
                            }
                            Text(clip.heading, style = VoiidFont.rounded(14), color = Color.White)
                            Text(
                                clip.caption, style = VoiidFont.rounded(12), color = Color.White.copy(alpha = 0.85f),
                                textDecoration = TextDecoration.Underline,
                                modifier = Modifier.clickable { haptics.tap(); showComments = true },
                            )
                        }
                        Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(24.dp)) {
                            ClipAction(
                                icon = if (liked) Icons.Default.Favorite else Icons.Outlined.FavoriteBorder,
                                label = "${clip.likes + if (liked) 1 else 0}",
                                tint = if (liked) VoiidColor.error else Color.White,
                            ) { haptics.tap(); liked = !liked; if (liked) clips.toggleLike(clip) }
                            ClipAction(Icons.AutoMirrored.Outlined.Chat, "${clip.comments}", Color.White) { haptics.tap(); showComments = true }
                            ClipAction(Icons.AutoMirrored.Filled.Send, "Share", Color.White) {}
                        }
                    }
                }
            }

            // Comments panel slides up below the shrunken reel.
            AnimatedVisibility(visible = showComments) {
                CommentsPanel(
                    clips = clips,
                    draft = commentDraft,
                    onDraftChange = { commentDraft = it },
                    onClose = { showComments = false },
                    onSend = {
                        val t = commentDraft.trim()
                        if (t.isNotEmpty()) { haptics.tap(); clips.addComment(t); commentDraft = "" }
                    },
                    modifier = Modifier.fillMaxWidth().height(totalH * 0.58f),
                )
            }
        }
    }
}

@Composable
private fun CommentsPanel(
    clips: ClipsStore,
    draft: String,
    onDraftChange: (String) -> Unit,
    onClose: () -> Unit,
    onSend: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier.clip(RoundedCornerShape(topStart = 20.dp, topEnd = 20.dp)).background(VoiidColor.background),
    ) {
        Box(
            Modifier.fillMaxWidth().padding(vertical = 8.dp), contentAlignment = Alignment.Center,
        ) { Box(Modifier.width(40.dp).height(4.dp).clip(CircleShape).background(VoiidColor.divider)) }
        Row(
            Modifier.fillMaxWidth().padding(horizontal = 24.dp).padding(bottom = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text("${clips.comments.size} comments", style = VoiidFont.rounded(15, FontWeight.SemiBold), color = VoiidColor.textPrimary)
            Spacer(Modifier.weight(1f))
            Icon(Icons.Default.Close, "Close", tint = VoiidColor.textSecondary, modifier = Modifier.size(20.dp).clickable { onClose() })
        }
        HorizontalDivider(color = VoiidColor.divider.copy(alpha = 0.5f))

        LazyColumn(Modifier.fillMaxWidth().weight(1f), contentPadding = androidx.compose.foundation.layout.PaddingValues(24.dp)) {
            items(clips.comments, key = { it.id }) { c -> CommentRow(c) }
        }

        HorizontalDivider(color = VoiidColor.divider.copy(alpha = 0.5f))
        Row(
            Modifier.fillMaxWidth().navigationBarsPadding().padding(horizontal = 16.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Row(
                Modifier.weight(1f).height(44.dp).clip(CircleShape).background(VoiidColor.fieldFill).padding(horizontal = 16.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Box(Modifier.weight(1f), contentAlignment = Alignment.CenterStart) {
                    if (draft.isEmpty()) Text("Add a comment…", style = VoiidFont.rounded(15), color = VoiidColor.placeholder)
                    BasicTextField(
                        value = draft, onValueChange = onDraftChange, singleLine = true,
                        textStyle = VoiidFont.rounded(15).merge(TextStyle(color = VoiidColor.textPrimary)),
                        cursorBrush = SolidColor(VoiidColor.primary), modifier = Modifier.fillMaxWidth(),
                    )
                }
            }
            Icon(Icons.AutoMirrored.Filled.Send, "Send", tint = VoiidColor.primary, modifier = Modifier.size(22.dp).clickable { onSend() })
        }
    }
}

@Composable
private fun CommentRow(c: VClipComment) {
    Row(
        Modifier.fillMaxWidth().padding(bottom = 16.dp),
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        VoiidAvatar(size = 32.dp, modifier = Modifier.clip(CircleShape))
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text(c.authorName, style = VoiidFont.rounded(13, FontWeight.SemiBold), color = VoiidColor.textPrimary)
            Text(c.text, style = VoiidFont.rounded(14), color = VoiidColor.textPrimary)
        }
        Icon(Icons.Outlined.FavoriteBorder, null, tint = VoiidColor.textSecondary, modifier = Modifier.size(13.dp))
    }
}

@Composable
private fun ClipAction(icon: ImageVector, label: String, tint: Color, onClick: () -> Unit) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(4.dp),
        modifier = Modifier.clickable(
            interactionSource = remember { MutableInteractionSource() }, indication = null,
        ) { onClick() },
    ) {
        Icon(icon, null, tint = tint, modifier = Modifier.size(28.dp))
        Text(label, style = VoiidFont.caption, color = Color.White)
    }
}
